content/2015.08.notifications-preamble.md

```diff
@@ -0,0 +1,166 @@
+
+* We should strive to be as *agnostic* as possible, meaning that locking into a particular service, back-end, or even programming language, should be avoided where possible.
+* Since we're web-centric, we should capitalise on HTTP wherever possible as well.
+* The notifications mechanism must be able to scale with a minimum of hardship; this applies to any eventual queues, workers, and so forth.
+* Finally, it must be robust - we can't be losing messages now, can we?
+
+# How?
+
+There are about as many ways to go about solving the problem as there are grains of sand on a beach (warning: previous statement may not be scientifically valid). The best way to determine the winning approach is simple: through single combat, [Thunderdome-style](https://duckduckgo.com/?q=beyond+thunderdome&ia=about). Unfortunately our manager put the kibosh on that plan for liability reasons, so instead we did the next best thing: we drew on a whiteboard.
+
+In the interest of absolutely transparency, here are some of the models that we worked through and our general observations on each.
+
+## The Simple
+
+One approach is to be as [hands-off](<https://i.imgur.com/xZQitlj.gif) as possible. The premise here is straightforward enough: Kinto drops notifications into a queue (like Redis or RabbitMQ). That's basically it.
+
+{% graphviz
+    dot {
+        digraph simple {
+            rankdir=LR;
+            Kinto [shape=box];
+            Queue [shape=box];
+            Consumer [shape=box];
+            Kinto -> Queue;
+            Queue -> Consumer [style=dotted];
+            Consumer -> Queue [style=dotted];
+        }
+    }
+%}
+
+Pros:
+
+* This model implies a highly limited scope insofar as Kinto is directly concerned; we would basically treat Kinto like a lightweight transport mechanism, and the queue as a fire-and-forget target.
+* The amount of additional code would be relatively small - in essence, only just enough to speak to a queue service and nothing more.
+* We could ship with support for a handful of external queuing services.
+
+Cons:
+
+* We'd have to code and maintain support for a handful of external queuing services.
+* It's ultimately up to the final consumer (receiver, destination, whatever) of the message to deal with the queue. In other words, if the queue is RabbitMQ, then the consumer needs to able to speak RabbitMQ too.
+* Kinto has no control, understanding, or insight into what happens once the message is handed off to the queuing service.
+
+The simplicity of this model, while attractive, has some fatal flaws. Chiefly, it runs contrary to our guideline of tech agnosticism as it would impose a particular external service dependency on the *consumer*. Furthermore, this model flirts with danger in terms of robustness - the fire-and-forget model isn't exactly optimised for reliability.
```

Comment:
I am actually not sure this is true: some queuing mechanisms have an ability to check if the message was delivered or not.